contracts/src/registry/EnhancedAccessControl.sol

```diff
@@ -0,0 +1,220 @@
+ * - Removing all assignments of a given role from a context.
+ */
+abstract contract EnhancedAccessControl is Context, ERC165 {
+    error EnhancedAccessControlUnauthorizedAccount(bytes32 context, bytes32 role, address account);
+    error EnhancedAccessControlBadConfirmation();
+
+    event EnhancedAccessControlRoleAdminChanged(bytes32 role, bytes32 previousAdminRole, bytes32 newAdminRole);
+    event EnhancedAccessControlRoleGranted(bytes32 context, bytes32 role, address account, address sender);
+    event EnhancedAccessControlRoleRevoked(bytes32 context, bytes32 role, address account, address sender);
+
+    /** @dev user role within a context. */
+    struct RoleAssignment {
+        bool hasRole;      // Indicates if the user has the role
+        uint256 version;   // Version tying this assignment to the current role version - see _roleVersion below
+    }
+
+    /** 
+     * @dev admin role that controls a given role. 
+     * Role -> AdminRole
+     */
+    mapping(bytes32 role => bytes32 adminRole) private _adminRoles;
+
+    /**
+     * @dev user role within a context.
+     * Context -> Role -> User -> RoleAssignment
+     */
+    mapping(bytes32 context => mapping(bytes32 role => mapping(address account => RoleAssignment roleAssignment))) private _roles;
+
+    /**
+     * @dev We use a version number to track the changes to the role assignments.
+     * This means we can easily remove all assignments of a given role without needing a loop.
+     */
+    mapping(bytes32 context => mapping(bytes32 role => uint256 version)) private _roleVersion;
+
+    /**
+     * @dev The root context.
+     */
+    bytes32 public constant ROOT_CONTEXT = bytes32(0);
+
+    /**
+     * @dev The default admin role.
+     */
+    bytes32 public constant DEFAULT_ADMIN_ROLE = 0x00;
+
+    /**
+     * @dev Modifier that checks that sender has a specific role within the given context. 
+     * Reverts with an {AccessControlUnauthorizedAccount} error including the required role.
+     */
+    modifier onlyRole(bytes32 context, bytes32 role) {
+        _checkRole(context, role);
+        _;
+    }
+
+    /**
+     * @dev See {IERC165-supportsInterface}.
+     */
+    function supportsInterface(bytes4 interfaceId) public view virtual override returns (bool) {
+        return interfaceId == type(EnhancedAccessControl).interfaceId || super.supportsInterface(interfaceId);
+    }
+
+    /**
+     * @dev Returns `true` if `account` has been granted `role`.
+     */
+    function hasRole(bytes32 context, bytes32 role, address account) public view virtual returns (bool) {
+        uint256 roleVersion = _roleVersion[context][role];
+        uint256 roleVersionInRoot = _roleVersion[ROOT_CONTEXT][role];
+
+        return (_roles[context][role][account].hasRole && _roles[context][role][account].version == roleVersion) ||
+               (_roles[ROOT_CONTEXT][role][account].hasRole && _roles[ROOT_CONTEXT][role][account].version == roleVersionInRoot);
```

Comment:
This could be made more efficient by only checking the `roleVersionInRoot` mapping if the check for a direct role fails.